File: src/flourish/analysis/talents/convoke.py
from flourish.analysis.attributor import TalentAttributor
from flourish.models.events import HealEvent, CastEvent, ApplyBuffEvent
from flourish.tracking.hot_tracker import HotTracker
from flourish.tracking.buff_tracker import BuffTracker

CONVOKE_SPELL_IDS = {391528, 323764}  # talent-tree ID + legacy (SL covenant) ID
CONVOKE_DURATION_MS = 4000
CONVOKE_DURATION_CG_MS = 3000  # with Cenarius' Guidance (-25%)
CENARIUS_GUIDANCE_NODE = 82063
DEFAULT_HEALING_RATIO = 0.7
CONVOKE_TAG = "convoke"


class ConvokeAttributor(TalentAttributor):
    """Convoke the Spirits: 4 sec channel firing 16 spells.
    Attributes healing from spells actually cast during the channel:
    - HoTs applied during channel are tagged; their ticks are attributed.
    - Direct heals during channel (no tracked HoT) are attributed.
    Pre-existing HoT ticks during the channel are NOT attributed.
    Duration is reduced to 3s with Cenarius' Guidance."""

    name = "Convoke the Spirits"
    talent_node_id = 82064
    talent_id = 103119  # WCL entryId (choice node vs Tree of Life)

    def __init__(self, healing_ratio: float = DEFAULT_HEALING_RATIO):
        super().__init__()
        self._channel_end = 0
        self._healing_ratio = healing_ratio

    def _channel_duration(self) -> int:
        if self.has_talent(CENARIUS_GUIDANCE_NODE):
            return CONVOKE_DURATION_CG_MS
        return CONVOKE_DURATION_MS

    def _is_channeling(self, timestamp: int) -> bool:
        return self._channel_end > 0 and timestamp <= self._channel_end

    def process_event(self, event, hot_tracker: HotTracker, buff_tracker: BuffTracker):
        if isinstance(event, CastEvent) and event.ability_id in CONVOKE_SPELL_IDS:
            self._channel_end = event.timestamp + self._channel_duration()

        # Tag HoTs applied during Convoke channel
        if self._is_channeling(event.timestamp) and isinstance(event, ApplyBuffEvent):
            hot = hot_tracker.get(event.target_id, event.ability_id)
            if hot:
                hot.tags.add(CONVOKE_TAG)

    def process_heal(self, event: HealEvent, hot_tracker: HotTracker, buff_tracker: BuffTracker) -> float:
        hot = hot_tracker.get(event.target_id, event.ability_id)

        # Convoke-tagged HoT tick (during or after channel)
        if hot and CONVOKE_TAG in hot.tags:
            return event.amount * self._healing_ratio

        # Direct heal during channel (no tracked HoT, e.g. Swiftmend)
        if self._is_channeling(event.timestamp) and hot is None:
            return event.amount * self._healing_ratio

        return 0.0

File: go-backend/internal/talents/convoke.go
package talents

import (
	"github.com/rdruid-talent-analyzer/go-backend/internal/models"
	"github.com/rdruid-talent-analyzer/go-backend/internal/tracking"
)

const (
	convokeDurationMS    = 4000
	convokeDurationCGMS  = 3000
	cenariusGuidanceNode = 82063
	convokeTag           = "convoke"
	convokeSkipCount     = 3 // first N rejuv/regrowth casts are opportunity cost
)

type ConvokeAttributor struct {
	BaseAttributor
	channelEnd     int
	skippableCasts int // rejuv/regrowth casts seen during current channel
}

func NewConvokeAttributor() *ConvokeAttributor {
	return &ConvokeAttributor{
		BaseAttributor: NewBaseAttributor("Convoke the Spirits", intPtr(82064), intPtr(103119)),
	}
}

func (a *ConvokeAttributor) channelDuration() int {
	if a.HasTalent(cenariusGuidanceNode) {
		return convokeDurationCGMS
	}
	return convokeDurationMS
}

func (a *ConvokeAttributor) isChanneling(ts int) bool {
	return a.channelEnd > 0 && ts <= a.channelEnd
}

func isSkippableConvokeSpell(abilityID int) bool {
	return abilityID == Rejuvenation || abilityID == GerminationRejuv || abilityID == Regrowth
}

func (a *ConvokeAttributor) ProcessEvent(event models.Event, hot *tracking.HotTracker, buff *tracking.BuffTracker) {
	if ce, ok := event.(*models.CastEvent); ok {
		if ce.AbilityID == ConvokeTheSpirits || ce.AbilityID == ConvokeLegacy {
			a.channelEnd = ce.Timestamp + a.channelDuration()
			a.skippableCasts = 0
		}
	}

	if ab, ok := event.(*models.ApplyBuffEvent); ok {
		if a.isChanneling(ab.Timestamp) {
			// Skip first N rejuv/regrowth applications (opportunity cost)
			if isSkippableConvokeSpell(ab.AbilityID) {
				a.skippableCasts++
				if a.skippableCasts <= convokeSkipCount {
					return
				}
			}
			h := hot.Get(ab.TargetID, ab.AbilityID)
			if h != nil {
				h.Tags[convokeTag] = true
			}
		}
	}
}

func (a *ConvokeAttributor) ProcessHeal(event *models.HealEvent, hot *tracking.HotTracker, buff *tracking.BuffTracker) float64 {
	h := hot.Get(event.TargetID, event.AbilityID)

	// Tagged HoT tick — this HoT was applied by Convoke (and not skipped)
	if h != nil && h.Tags[convokeTag] {
		return float64(event.Amount)
	}

	// Direct heal during channel (no HoT, e.g. Swiftmend, Wild Growth initial)
	if a.isChanneling(event.Timestamp) && h == nil {
		// Skip skippable direct heals that correspond to skipped casts
		if isSkippableConvokeSpell(event.AbilityID) {
			return 0
		}
		return float64(event.Amount)
	}

	return 0
}
